docs/mpirun.rst

```diff
@@ -0,0 +1,153 @@
+``-map-by slot`` allows you to have a mixture of different NUMA configurations because the default behavior is to bind
+to the socket.
+
+``-mca pml ob1`` and ``-mca btl ^openib`` flags force the use of TCP for MPI communication.  This avoids many
+multiprocessing issues that Open MPI has with RDMA which typically result in segmentation faults.  Using TCP for MPI
+does not have noticeable performance impact since most of the heavy communication is done by NCCL, which will use RDMA
+via RoCE or InfiniBand if they're available (see `Horovod on GPU <gpus.md>`_).  Notable exceptions from this rule are
+models that heavily use ``hvd.broadcast()`` and ``hvd.allgather()`` operations.  To make those operations use RDMA,
+read the `Open MPI with RDMA <#open-mpi-with-rdma>`_ section below.
+
+With the ``-x`` option you can specify (``-x NCCL_DEBUG=INFO``) or copy (``-x LD_LIBRARY_PATH``) an environment variable to
+all the workers.
+
+Custom SSH ports
+----------------
+
+Custom SSH ports can be specified by ``-mca plm_rsh_args "-p <port>"`` as follows:
+
+.. code-block:: bash
+
+    mpirun -np 16 \
+        -H server1:4,server2:4,server3:4,server4:4 \
+        -bind-to none -map-by slot \
+        -mca plm_rsh_args "-p 12345"
+        -x NCCL_DEBUG=INFO -x LD_LIBRARY_PATH -x PATH \
+        -mca pml ob1 -mca btl ^openib \
+        python train.py
+
+This is frequently useful in the case of `running Horovod in Docker environment <docker.md>`_.
+
+Open MPI with RDMA
+------------------
+
+As noted above, using TCP for MPI communication does not have any significant effects on performance in the majority of
+cases. Models that make heavy use of ``hvd.broadcast()`` and ``hvd.allgather()`` operations are exceptions to that rule.
+
+Default Open MPI ``openib`` BTL that provides RDMA functionality does not work well with MPI multithreading.  In order
+to use RDMA with ``openib``, multithreading must be disabled via ``-x HOROVOD_MPI_THREADS_DISABLE=1`` option.  See the
+example below:
+
+.. code-block:: bash
+
+    mpirun -np 16 \
+        -H server1:4,server2:4,server3:4,server4:4 \
+        -bind-to none -map-by slot \
+        -x NCCL_DEBUG=INFO -x LD_LIBRARY_PATH -x HOROVOD_MPI_THREADS_DISABLE=1 -x PATH \
+        -mca pml ob1 \
+        python train.py
+
+Other MPI RDMA implementations may or may not benefit from disabling multithreading, so please consult vendor
+documentation.
+
+Hangs due to non-routed network interfaces
+------------------------------------------
+
+Having network interfaces that are not routed can cause Open MPI to hang. An example of such interface is ``docker0``.
+
+If you see non-routed interfaces (like ``docker0``) in the output of ``ifconfig``, you should tell Open MPI and NCCL to not
+use them via the ``-mca btl_tcp_if_exclude <interface>[,<interface>]`` and ``NCCL_SOCKET_IFNAME=^<interface>[,<interface>]``
+parameters.
+
+.. code-block:: bash
+
+    ifconfig
+
+Produces output like this::
```

Comment:
you might need to add a space after "this"

Reply:
AFAIK space is not necessary - http://docutils.sourceforge.net/docs/user/rst/quickstart.html#preformatting-code-samples